server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletGoalState.java

```diff
@@ -122,39 +123,41 @@ public static TabletGoalState compute(TabletMetadata tm, TabletState currentStat
             "Expected current tablet location %s %s", tm.getExtent(), tm.getLocation());
         var tsii = new TabletServerIdImpl(tm.getLocation().getServerInstance());
 
-        var resourceGroup = params.getResourceGroup(tm.getLocation().getServerInstance());
-
-        if (resourceGroup != null) {
-          var reassign = balancer.needsReassignment(new TabletBalancer.CurrentAssignment() {
-            @Override
-            public TabletId getTablet() {
-              return new TabletIdImpl(tm.getExtent());
-            }
+        String resourceGroup = params.getResourceGroup(tm.getLocation().getServerInstance());
+
+        if (resourceGroup == null || resourceGroup.isBlank()) {
+          // The current tserver for the tablet should be in the TabletManagementParams object
+          // which is derived from the LiveTServerSet in the Manager. Not existing here could
+          // either be an issue with the ServiceLock, where the group is not set, or a bug in
+          // the code between the LiveTServerSet and here. Use the default resource group in
+          // this case for determining if the tablet needs to be re-assigned.
+          log.warn(
+              "Resource group for TabletServer {} is not set, using the default resource"
+                  + "group. If this issue persists please submit an issue.",
+              tm.getLocation().getServerInstance(), tm.getExtent());
+          resourceGroup = Constants.DEFAULT_RESOURCE_GROUP_NAME;
```

Comment:
I don't think this should fall back to the default resource group in this case, because if we get it here its either caused by a bug or race condition.  In either of those cases I do not think its good to start assigning stuff to the default resource group.  If it is a race condition the later passes of the TGW should see the correct data and be able to make the right decision.

Digging into the code I found the following. 

* Tto get this point in the code the current state of the tablet has to be HOSTED  
* For the current state to be HOSTED that means TabletState.compute() saw the tablet server in the set of live tservers.

So if we are seeing a tablet that is in the set of live tsevers and it does not have a RG then its either a bug or race condition.  TabletState.compute() and TabletGoalState.compute() are only called in two places and I think a race condition could occur in one of those in TGW having just looked at the code and how it calls those two functions.

Would probably be better to adjust the existing codes logging of an error message to debug for now since there is a known race condition, but not use the default RG. We could open a follow on issue to try to figure out what is going on when we see this.  We could reread the data and see if the live tserver set and RG are still in disagreement, if that is the case it points to a bug and should log an error.  If after rereading the data we see that things are in agreement we could log a debug as this indicates a race condition.

Also if we see a blank RG I think we should throw an exception as this seems like a bug and not something a race condition would cause.  I am assuming a server can not be started with a blank RG, so should not silently ignore that here.

Reply:
> Would probably be better to adjust the existing codes logging of an error message to debug for now since there is a known race condition

OK, I'll modify this PR to do that.

> Also if we see a blank RG I think we should throw an exception as this seems like a bug and not something a race condition would cause. I am assuming a server can not be started with a blank RG, so should not silently ignore that here.

I believe that is correct. I looked at the ServiceLock code yesterday and I'm pretty sure that the resource group is mandatory when that object is created. So a cause for a blank RG would be:

  1. a change to service lock such that the RG is not being serialized to ZK, or
  2. a bug between the reading of ZK and this point in the code

Reply:
@keith-turner - I changed the log message in f36120c. I didn't change to `debug` and used `warn` instead.

Reply:
Warn is good, we can adjust it down later if its too frequent.